chapters/os_tooling.md

```diff
@@ -0,0 +1,85 @@
+* [clspv](https://github.com/google/clspv) compiler and [clvk](https://github.com/kpet/clvk) runtime layer enable OpenCL applications to be executed with Vulkan drivers.
+* [SPIR-V Tools](https://github.com/KhronosGroup/SPIRV-Tools) provide a set of utilities to process SPIR-V binaries including __spirv-opt__ optimizer, __spirv-link__ linker, __spirv-dis__/__spirv-as__ (dis-)assembler, and __spirv-val__ validator.
+
+### Compiling Kernels to SPIR-V
+
+The open source tools can be used to perform full compilation from OpenCL kernel sources into SPIR-V.
+
+<p align="center">
+<br>
+<img src="../images/opencl_to_spirv_tooling.jpg">
+<br> <br>
+  <b>Offline compilation flow for OpenCL kernels into SPIR-V</b>
+<br> <br>
+</p>
+
+When tools are used to target OpenCL drivers they operate on OpenCL flavor of SPIR-V and when tools target Vulkan drivers they operate on Vulkan flavor of SPIR-V instead. Some tools can ingest both SPIR-V dialects. However, it is generally not possible to transform or mix the modules of different SPIR-V flavors. For example, it is generally not possible to use __spirv-link__ to link the modules in both OpenCL and Vulkan formats of SPIR-V.
+
+#### Examples
+
+If you want to try the above compilation flow for yourself, after installing the tools, you can use the following commands.
+
+##### Compile for OpenCL runtime
+
+__(i)__ Compiling OpenCL C/C++ for OpenCL file into SPIR flavor LLVM IR (for 32 bit targets) formats:
+
+```
+clang -c -target spir -O0 -emit-llvm -o test.bc test.cl
+```
+Using `-cl-std` changes the language version compiled for.
+
+To compile C++ for OpenCL source pass `-cl-std=CLC++` or use the file extension `.clcpp`.
+
+```
+clang -cl-std=CLC++ -c -target spir -O0 -emit-llvm -o test.bc test.cl
+clang -c -target spir -O0 -emit-llvm -o test.bc test.clcpp
+```
+If debugging support is needed then the `-g` flag can be used in the clang invocation:
+
+```
+clang -c test.cl -target spir -o test.bc -g
+```
+
+__Note:__ In clang releases up to 12.0, calling most builtin functions requires the extra flags `-Xclang -finclude-default-header` to be passed to clang. Refer to the release documentation for more details.
+
+__(ii)__ Converting LLVM IR into SPIR-V. 
+
+```
+llvm-spirv test.bc -o test.spv
+```
+
+__Note:__ Converting IR produced with optimization levels other than `-O0` is only available as an experimental feature and it is not guaranteed to work. The majority of common functionality is expected to convert even when optimizations are enabled. Developers are encouraged to file a bug report when issues are encountered. When translating modules obtained with optimizations can not be not supported, the stand-alone __spirv-opt__ tool can be used to optimize in SPIR-V format once the conversion is done from LLVM IR produced with optimizations disabled.
```

Comment:
Looks good, just rewording a little to make it clearer for the reader:

```suggestion
__Note:__ Converting IR produced with optimization levels other than `-O0` is only available as an experimental feature and is not guaranteed to work. The conversion should work for the majority of IR constructs even when optimizations are enabled, but certain IR constructs may cause issues. Developers are encouraged to file a bug report when issues are encountered. As a workaround when encountering an issue in translating modules obtained with optimizations, generate LLVM IR with optimizations disabled, and then use the stand-alone __spirv-opt__ tool to optimize at the SPIR-V level.
```

Reply:
For the developers I prefer not to tight the description to the IR constructs since they are not supposed to know details about it.

Reply:
```suggestion
__Note:__ Converting IR produced with optimization levels other than `-O0` is only available as an experimental feature and it is not guaranteed to work. The majority of common functionality is expected to convert even when optimizations are enabled. Developers are encouraged to file a bug report when issues are encountered. As a workaround when encountering an issue in translating modules obtained with optimizations, generate LLVM IR with optimizations disabled, and then use the stand-alone __spirv-opt__ tool to optimize at the SPIR-V level.
```

Reply:
How about we keep the first part as is?

Reply:
The words "common functionality" in combination with "is expected to convert" is unclear to me.  Maybe I misunderstood what you are trying to say with "common functionality" in this context?

Regardless, I think you should still rephrase "expected to convert" into "conversion [...] is expected to succeed".

Reply:
Ok, I have changed to

` In the majority of cases, the conversion is expected to succeed when optimizations are enabled.`

Does it make sense?

Reply:
Yes, that's better, thanks!